engine/src/types.rs

```diff
@@ -185,6 +212,63 @@ macro_rules! declare_types {
     };
 }
 
+// type Map<'a> = HashMap<&'a str, LhsValue<'a>>;
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
+pub struct Map<'a> {
```

Comment:
Instead of having to duplicate same type in each item, this should probably be defined in types.rs similarly to `RhsValues` (which is conceptually very similar - a collection of values with a shared tag).

Reply:
Well `RhsValues` is not generic enough; It just handles a collection of primary type whereas in case of map/array we want all the values to be of the same arbitrary type:
* Map of `<something>`
* Array of `<something>`

Reply:
I'm not sure I understand the difference. Both sound generic over some type `T`, and can hold only uniform elements, no?

Reply:
To my understanding `RhsValues` is not generic. It can only holds list of primary type:
```Rust
enum RhsValues {
    Ip(Vec<IpRange>),
    Bytes(Vec<Bytes>),
    Int(Vec<RangeInclusive<i32>>),
    Bool(Vec<UninhabitedBool>),
}
```
Its a pre-defined fixed list of possibilities.
I don't get how would you dynamically define a `Map of Array of Bytes` ?

Reply:
Uh... we're not going to limit nesting in any way, allowing users to combine types in whatever complex structures they want?

Reply:
IMO, i don't think such limitation is valuable. One should be able to define an arbitrary complex schema without the need to modify the core engine because some types combinations are missing.
Plus, it's not that complicated to introduce as shown in this PR.